test/test_heuristic.cpp

```diff
@@ -12,64 +12,935 @@
+ *
+ * @param nodes vector of all nodes (each at the position corresponding to its
+ * id)
+ * @param nodeId id of the node from which to find the path to the root
+ */
+std::vector<std::size_t>
+getPathToRoot(std::vector<HeuristicMapper::Node>& nodes, std::size_t nodeId) {
+  std::vector<std::size_t> path{};
+  if (nodeId >= nodes.size() || nodes[nodeId].id != nodeId) {
+    throw std::runtime_error("Invalid node id " + std::to_string(nodeId));
+  }
+  auto* node = &nodes[nodeId];
+  while (node->parent != node->id) {
+    path.push_back(node->id);
+    if (node->parent >= nodes.size() ||
+        nodes[node->parent].id != node->parent) {
+      throw std::runtime_error("Invalid parent id " +
+                               std::to_string(node->parent) + " for node " +
+                               std::to_string(node->id));
+    }
+    node = &nodes[node->parent];
+  }
+  path.push_back(node->id);
+  return path;
+}
+
+class InternalsTest : public HeuristicMapper, public testing::Test {
+protected:
+  // NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
+  static Architecture defaultArch;
+
+  InternalsTest() : HeuristicMapper(qc::QuantumComputation{1}, defaultArch) {}
+  void SetUp() override { results = MappingResults{}; }
+};
 
-TEST(Functionality, NodeCostCalculation) {
-  const double                  tolerance = 1e-6;
-  const CouplingMap             cm        = {{0, 1}, {1, 2}, {3, 1}, {4, 3}};
-  Architecture                  arch{5, cm};
-  const SingleQubitMultiplicity empty1Mult                  = {};
-  const std::unordered_set<std::uint16_t>& consideredQubits = {0, 1, 2, 3, 5};
-  const TwoQubitMultiplicity               multiplicity     = {{{0, 1}, {5, 2}},
-                                                               {{2, 3}, {0, 1}}};
-  const std::array<std::int16_t, MAX_DEVICE_QUBITS> qubits  = {4, 3, 1, 2, 0};
-  const std::array<std::int16_t, MAX_DEVICE_QUBITS> locations = {4, 2, 3, 1, 0};
-
-  const std::vector<std::vector<Exchange>> swaps = {
-      {Exchange(0, 1, qc::OpType::Teleportation)},
-      {Exchange(1, 2, qc::OpType::SWAP)}};
-
-  HeuristicMapper::Node node(0, 0, qubits, locations, swaps, 5., 0, false,
-                             false);
-  node.updateHeuristicCost(arch, empty1Mult, multiplicity, consideredQubits);
+// NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
+Architecture InternalsTest::defaultArch{1, {}};
+
+TEST_F(InternalsTest, NodeCostCalculation) {
+  results.config.heuristic          = Heuristic::GateCountMaxDistance;
+  results.config.lookaheadHeuristic = LookaheadHeuristic::None;
+  results.config.layering           = Layering::Disjoint2qBlocks;
+
+  architecture->loadCouplingMap(5, {{0, 1}, {1, 2}, {3, 1}, {4, 3}});
+  qc = qc::QuantumComputation{5};
+  // layer 0 distances:
+  // 0-1: 2 swaps & 2 reversals
+  // 2-3: 0 swaps & 1 reversals
+  qc.cx(qc::Control{0}, 1);
+  qc.cx(qc::Control{0}, 1);
+  qc.cx(qc::Control{0}, 1);
+  qc.cx(qc::Control{0}, 1);
```

Comment:
Just for reference. I think you should be able to get away with `qc.cx(0,1)` (or, potentially `qc.cx(0U, 1U)`)

Reply:
Hm true, I could have sworn I had been running into compiler errors/warnings without the explicit conversion, but it seems to work now. Great, thanks!